mne_connectivity/spectral/time.py

```diff
@@ -457,135 +472,89 @@ def _spectral_connectivity(data, method, kernel, foi_idx,
-        s_xy = _compute_csd(w[:, w_y], w[:, w_x], weights)
+        dphi = s_xy / np.abs(s_xy)
+        dphi = _smooth_spectra(dphi, kernel)
         s_xy = _smooth_spectra(s_xy, kernel)
-        out = np.abs(np.mean(np.sign(np.imag(s_xy)),
-                             axis=-1, keepdims=True))
-        # mean inside frequency sliding window (if needed)
-        if isinstance(foi_idx, np.ndarray) and faverage:
-            return _foi_average(out, foi_idx)
-        else:
-            return out
-
-    # define the function to compute in parallel
-    parallel, p_fun, n_jobs = parallel_func(
-        pairwise_pli, n_jobs=n_jobs, verbose=verbose, total=total)
-
-    return parallel(p_fun(s, t) for s, t in zip(source_idx, target_idx))
-
-
-def _wpli(w, kernel, foi_idx, source_idx, target_idx, n_jobs, verbose, total,
-          faverage, weights):
-    """Pairwise weighted phase-lag index.
-
-    Input signal w is of shape (n_epochs, n_chans, n_tapers, n_freqs,
-    n_times)."""
-    def pairwise_wpli(w_x, w_y):
-        s_xy = _compute_csd(w[:, w_y], w[:, w_x], weights)
-        s_xy = _smooth_spectra(s_xy, kernel)
-        con_num = np.abs(s_xy.imag.mean(axis=-1, keepdims=True))
-        con_den = np.mean(np.abs(s_xy.imag), axis=-1, keepdims=True)
-        out = con_num / con_den
-        # mean inside frequency sliding window (if needed)
-        if isinstance(foi_idx, np.ndarray) and faverage:
-            return _foi_average(out, foi_idx)
-        else:
-            return out
-
-    # define the function to compute in parallel
-    parallel, p_fun, n_jobs = parallel_func(
-        pairwise_wpli, n_jobs=n_jobs, verbose=verbose, total=total)
+        out = []
+        for m in method:
+            if m == 'coh':
+                s_xx = psd[w_x]
+                s_yy = psd[w_y]
+                coh = np.abs(s_xy.mean(axis=-1, keepdims=True)) / \
+                    np.sqrt(s_xx.mean(axis=-1, keepdims=True) *
+                            s_yy.mean(axis=-1, keepdims=True))
+                out.append(coh)
+
+            if m == 'plv':
+                dphi_mean = dphi.mean(axis=-1, keepdims=True)
+                plv = np.abs(dphi_mean)
+                out.append(plv)
+
+            if m == 'ciplv':
+                rplv = np.abs(np.mean(np.real(dphi), axis=-1, keepdims=True))
+                iplv = np.abs(np.mean(np.imag(dphi), axis=-1, keepdims=True))
+                ciplv = iplv / (np.sqrt(1 - rplv ** 2))
+                out.append(ciplv)
+
+            if m == 'pli':
+                pli = np.abs(np.mean(np.sign(np.imag(s_xy)),
+                                     axis=-1, keepdims=True))
+                out.append(pli)
+
+            if m == 'wpli':
+                con_num = np.abs(s_xy.imag.mean(axis=-1, keepdims=True))
+                con_den = np.mean(np.abs(s_xy.imag), axis=-1, keepdims=True)
+                wpli = con_num / con_den
+                out.append(wpli)
+
+            if m == 'cs':
+                out.append(s_xy)
```

Comment:
WDYT about keeping the basic pairwise "computation" functions as separate functions, so that way they can be self-documented in their docstring? E.g. keeping all the old functions?

And then to add say a new spectral-connectivity function, one would have to: i) add a new private function e.g. `def _newpli(x, y)` and then ii) update `_parallel_con` to add this option. 

Might be easier to read. But also not 100% necessary.

Reply:
It's an option for sure. However, it should be done such that we avoid computing the cross-spectral density multiple times. This is slow due to the potentially large size of the arrays according to some profiling I did.

Reply:
Will take a look to see if this is possible then. 

I wonder if it's possible to keep even the same "spectral-conn" base functions across 'epochs' and 'time' computation. I.e. there are the classes inside `spectral_conn_epochs.py` which also compute certain spectral connectivity. They seem to all operate on a pairwise nature, so perhaps we can even consolidate into one place, so then adding one connectivity function adds it for both 'epochs' and 'time'

Reply:
I'm afraid it wouldn't be practical to use the classes from `spectral_connectivity_epochs`. Those estimate connectivity as mean over epochs with the assumption that the connectivity measure is stationary within an Epoch. An estimator's `accumulate` method is therefore called only once per Epoch. Cross-spectral density is computed in a single time window spanning the whole Epoch (with the exception of `cwt_morlet` mode).

In `spectral_connectivity_time` the pairwise computation involves summing over the time axis and `multitaper` actually means convolution with a DPSS-tapered sinusoid (similar to Gaussian tapered sinusoid for `morlet`. Summing over the time axis using something like the `accumulate` function in `epochs` would necessitate a for-loop over time. There's also the possibility of fetching connectivity over time separately for each epoch, which of course doesn't exist in `epochs`. 

In my opinion, accommodating both within the same classes would probably lead to major refactoring of `spectral_connectivity_epochs` and potentially make the code less readable.

Sorry if the explanation is too basic, I'm mostly writing it down to keep it straight for myself.

Reply:
WDYT about doing it this way?

```python

def _parallel_con(w, method, kernel, foi_idx, source_idx, target_idx, n_jobs,
                  verbose, total, faverage, weights):
    """Compute spectral connectivity in parallel.

    Input signal w is of shape (n_chans, n_tapers, n_freqs, n_times)."""

    if 'coh' in method:
        # psd
        if weights is not None:
            psd = weights * w
            psd = psd * np.conj(psd)
            psd = psd.real.sum(axis=1)
            psd = psd * 2 / (weights * weights.conj()).real.sum(axis=0)
        else:
            psd = w.real ** 2 + w.imag ** 2
            psd = np.squeeze(psd, axis=1)

        # smooth
        psd = _smooth_spectra(psd, kernel)

    def pairwise_con(w_x, w_y):
        # csd
        if weights is not None:
            s_xy = np.sum(weights * w[w_x] * np.conj(weights * w[w_y]), axis=0)
            s_xy = s_xy * 2 / (weights * np.conj(weights)).real.sum(axis=0)
        else:
            s_xy = w[w_x] * np.conj(w[w_y])
            s_xy = np.squeeze(s_xy, axis=0)
        s_xy = _smooth_spectra(s_xy, kernel)
        out = []
        conn_func = {'plv': _plv, 'ciplv': _ciplv, 'pli': _pli, 'wpli': _wpli,
                     'coh': _coh, 'cs': _cs}
        for m in method:
            if m == 'coh':
                s_xx = psd[w_x]
                s_yy = psd[w_y]
                out.append(conn_func[m](s_xx, s_yy, s_xy))
            else:
                out.append(conn_func[m](s_xy))

        for i, _ in enumerate(out):
            # mean inside frequency sliding window (if needed)
            if isinstance(foi_idx, np.ndarray) and faverage:
                out[i] = _foi_average(out[i], foi_idx)
            # squeeze time dimension
            out[i] = out[i].squeeze(axis=-1)

        return out

    # only show progress if verbosity level is DEBUG
    if verbose != 'DEBUG' and verbose != 'debug' and verbose != 10:
        total = None

    # define the function to compute in parallel
    parallel, p_fun, n_jobs = parallel_func(
        pairwise_con, n_jobs=n_jobs, verbose=verbose, total=total)

    return parallel(p_fun(s, t) for s, t in zip(source_idx, target_idx))


def _plv(s_xy):
    s_xy = s_xy / np.abs(s_xy)
    plv = np.abs(s_xy.mean(axis=-1, keepdims=True))
    return plv


def _ciplv(s_xy):
    s_xy = s_xy / np.abs(s_xy)
    rplv = np.abs(np.mean(np.real(s_xy), axis=-1, keepdims=True))
    iplv = np.abs(np.mean(np.imag(s_xy), axis=-1, keepdims=True))
    ciplv = iplv / (np.sqrt(1 - rplv ** 2))
    return ciplv


def _pli(s_xy):
    pli = np.abs(np.mean(np.sign(np.imag(s_xy)),
                         axis=-1, keepdims=True))
    return pli


def _wpli(s_xy):
    con_num = np.abs(s_xy.imag.mean(axis=-1, keepdims=True))
    con_den = np.mean(np.abs(s_xy.imag), axis=-1, keepdims=True)
    wpli = con_num / con_den
    return wpli


def _coh(s_xx, s_yy, s_xy):
    con_num = np.abs(s_xy.mean(axis=-1, keepdims=True))
    con_den = np.sqrt(s_xx.mean(axis=-1, keepdims=True) *
                      s_yy.mean(axis=-1, keepdims=True))
    coh = con_num / con_den
    return coh


def _cs(s_xy):
    return s_xy.mean(axis=-1, keepdims=True)
```

Reply:
Yeah that looks better and more similar to the epochs file. 

Idr but can we put the pairwise function outside? Nested functions imo are harder to read. But if that breaks the paralleliation via joblib, then nvm.

Reply:
I'll try if the pairwise function works outside `_parallel_con`. Fully agree that nested functions are ugly.